### PR TITLE
feat: data-parity skill — TypeScript orchestrator, ClickHouse driver, partition support

### PR DESCRIPTION
**Pull request by suryaiyer95**
## Summary

Adds the full data-parity diffing pipeline for the \`data_diff\` LLM tool: a TypeScript orchestrator that bridges the Rust \`DataParitySession\` state machine with all database drivers, plus a comprehensive SKILL.md that guides the model through algorithm selection.

### Core: TypeScript orchestrator (\`data-diff.ts\`, +799 lines)

**Cooperative state machine bridge** — TypeScript executes SQL tasks generated by the Rust engine and feeds results back. Neither layer touches the other's domain.

- \`runDataDiff\` — top-level entry point; dispatches to single-table or partitioned diff
- \`runPartitionedDiff\` — runs independent diffs per partition with \`mergeOutcomes\`
- \`resolveTableSources\` + \`injectCte\` — compares arbitrary SQL queries by wrapping them as CTEs
- \`discoverExtraColumns\` — auto-discovers non-key columns with two-layer audit exclusion (name patterns + schema-level DEFAULT detection across 9 dialects)
- \`buildColumnDiscoverySQL\` — per-dialect INFORMATION_SCHEMA query (Postgres, Snowflake, BigQuery, MySQL, Redshift, SQLite, DuckDB, Trino/Presto, fallback)
- \`dateTruncExpr\` — dialect-aware date truncation for date-partition modes (BigQuery, MySQL/MariaDB, Oracle, Spark/Databricks, default \`DATE_TRUNC\`)
- \`buildPartitionDiscoverySQL\` / \`buildPartitionWhereClause\` — date, numeric, and categorical partition support

### New: ClickHouse driver (\`clickhouse.ts\`, +135 lines)

Pure HTTP driver using \`JSONCompactEachRowWithNamesAndTypes\` — no external package required. Handles DDL (empty response) vs SELECT (typed rows) branches correctly.

### \`ExecuteOptions\` — all 10+ drivers updated

New \`noLimit?: boolean\` flag bypasses the default 1,000-row cap for data-diff pipelines. Applied consistently across Postgres, Snowflake, BigQuery, Redshift, MySQL, SQLite, DuckDB, Trino, Databricks, ClickHouse.

### SKILL.md (+341 lines)

Comprehensive algorithm-selection guide:
- **CRITICAL**: \`joindiff\` cannot see a second table across warehouses — always use \`hashdiff\` or \`auto\` for cross-DB comparisons
- Decision tree: same-warehouse vs cross-warehouse, table vs query comparison
- Output format: facts only (counts, values, row samples) — no editorializing
- Full parameter reference for \`DataDiffParams\`

---

## Fixes applied (from code review)

- **Security** — \`buildColumnDiscoverySQL\` now escapes single quotes in all interpolated table name parts (\`esc()\`) — prevents SQL injection via crafted source/target table names
- **Bug** — \`dateTruncExpr\` Oracle case added: \`TRUNC(col, 'MONTH')\` — Oracle does not have \`DATE_TRUNC\`

## Test plan
- [x] Cross-DB comparison (pg_source vs pg_target) — agent uses \`hashdiff\` automatically
- [x] TPC-H migration validation — output is clean fact-reporting
- [x] SQL query comparison (same-warehouse) — \`joindiff\` used correctly
- [x] TypeScript typecheck — passing

## Summary by CodeRabbit

* **New Features**
  * Data Parity tool for table-vs-table and SQL-vs-SQL comparisons with profiling, column/row diffs, human-readable summaries, and registered runtime integration.
  * Partitioned diff execution with per-partition and aggregated summaries.
  * Automatic detection/exclusion of audit/timestamp-like columns and schema-driven extra-column discovery.

* **Documentation**
  * Added a comprehensive data parity skill guide and updated prompts to invoke the feature.

* **Tests**
  * Updated test mocks for manifest and SQL-translate responses.

* **Other**
  * Query execution gained a “no-limit” mode to avoid automatic truncation.

### CODE REVIEW

**Review by suryaiyer95 — COMMENTED** [review]
## Multi-Model Code Review — altimate-code#493

*Reviewed by 4-model panel: Claude Sonnet 4.6, Gemini 3.1 Pro, Grok 4, Kimi K2.5. Unanimous verdict: REQUEST CHANGES.*

> **Note:** The PR title significantly undersells scope. This contains a 799-line TypeScript orchestrator, new ClickHouse driver, \`ExecuteOptions\` refactored across 10+ drivers, column auto-discovery, partition support, and a 341-line SKILL.md.

**Verdict**: REQUEST CHANGES (fix MAJOR-1 before merge)
**Major Issues**: 1 | **Minor Issues**: 4 | **NITs**: 3

---

### Major Issues

**MAJOR-1 — SQL injection via table name interpolation in \`buildColumnDiscoverySQL\` (Security)**
- **Location**: \`packages/opencode/src/altimate/native/connections/data-diff.ts\` — \`buildColumnDiscoverySQL\`
- Table name parts are interpolated directly into SQL:
  ```typescript
  schemaFilter = `table_schema = '${parts[1]}'`
  tableFilter  = `table_name  = '${parts[2]}'`
  ```
  \`tableName\` flows from \`DataDiffParams.source\` which comes from the LLM/user. A crafted table name like \`orders' OR 1=1 --\` produces injected SQL.
- **Fix**:
  ```typescript
  const escape = (s: string) => s.replace(/'/g, "''")
  schemaFilter = `table_schema = '${escape(parts[1])}'`
  tableFilter  = `table_name  = '${escape(parts[2])}'`
  ```

---

### Minor Issues

**MINOR-1 — \`discoverExtraColumns\` silently falls back to key-only comparison (Logic Error)**
- When column discovery fails, the diff runs in key-only mode — reporting all rows as identical even when non-key values differ. No warning is surfaced.
- **Fix**: Include a \`warning\` field in \`DataDiffResult\` when key-only fallback occurs.

**MINOR-2 — \`dateTruncExpr\` missing Oracle dialect (Bug)**
- Oracle falls through to the default \`DATE_TRUNC\` case. Oracle uses \`TRUNC(col, 'MONTH')\` — date-partitioned diffs on Oracle tables will fail.
- **Fix**: Add \`case "oracle": return \`TRUNC(\${column}, '\${g.toUpperCase()}')\`\`

**MINOR-3 — Unqualified table names match across multiple schemas (Bug)**
- \`buildColumnDiscoverySQL\` with no schema filter returns columns from all schemas with that table name — producing incorrect \`extra_columns\`.
- **Fix**: Default \`schemaFilter\` to the connection's current schema, or document that fully-qualified names are required.

**MINOR-4 — \`MAX_STEPS = 200\` hardcoded and undocumented (Design)**
- Not configurable, not documented in \`DataDiffParams\` or tool description.
- **Fix**: Expose as optional \`max_steps?: number\` in \`DataDiffParams\`, defaulting to 200.

---

### NITs
- N1: \`mergeOutcomes\` caps \`diff_rows\` at 100 with no documentation — callers don't know the result is partial
- N2: \`partitionColumn\` identifier interpolated directly in \`buildPartitionDiscoverySQL\` — inconsistent with MAJOR-1 fix
- N3: \`effectiveLimit\` pattern repeated across 10+ drivers

---

### Positive Observations
- The cooperative state machine bridge (TypeScript orchestrator ↔ Rust engine) is the right architecture.
- Column auto-discovery with two-layer audit exclusion is thorough.
- \`ExecuteOptions.noLimit\` correctly applied across all 10+ drivers.
- SKILL.md \`CRITICAL: joindiff only sees one connection for cross-warehouse comparisons\` is exactly the right guardrail.
- ClickHouse driver using \`JSONCompactEachRowWithNamesAndTypes\` with pure HTTP is clean.
- Dynamic import gracefully handles NAPI unavailability — unblocks CI.

**Review by suryaiyer95 — COMMENTED** [review]
## Code Review: Data Parity TypeScript Orchestrator + ClickHouse Driver

**Verdict:** APPROVE with minor issues
**Reviewed by:** Claude (sole reviewer — external models unavailable)

### Major Issue (1)

#### 1. SQL Injection via `buildColumnDiscoverySQL` — ClickHouse/Snowflake paths (Security)

**Location:** `packages/opencode/src/altimate/native/connections/data-diff.ts`

The ClickHouse path uses `DESCRIBE TABLE ${tableName}` and Snowflake uses `SHOW COLUMNS IN TABLE ${tableName}` with raw string interpolation. While the `information_schema` paths properly escape with `esc()`, ClickHouse/Snowflake bypass escaping entirely.

A table name like `orders; DROP TABLE users --` would be injected verbatim. The `tableName` comes from LLM tool params, so prompt injection or user error could trigger this.

**Fix:** Apply identifier quoting (backtick for ClickHouse, double-quote for Snowflake) or validate table names against `[a-zA-Z0-9_.]`.

### Minor Issues (4)

#### 2. Profile outcome formatting uses wrong field names — `tools/data-diff.ts`

The TS formatter checks `col.source_stats`/`col.target_stats` but Rust serializes as `col.table1`/`col.table2`. Stats fields are `non_null_count`/`distinct_count` in Rust but the TS uses `count`/`null_count`. Profile output would show undefined values.

#### 3. CTE injection doesn't handle ClickHouse `SETTINGS` clause — `data-diff.ts:injectCte()`

JoinDiff for ClickHouse emits `... SETTINGS join_use_nulls = 1`. CTE-wrapped queries with `SETTINGS` may not propagate correctly in all ClickHouse versions.

#### 4. Partition discovery fails on query sources — `data-diff.ts:runPartitionedDiff()`

When source is a SQL query, `table1Name` resolves to `__diff_source` but the partition discovery SQL doesn't include the CTE prefix. This would fail with "table `__diff_source` not found".

#### 5. `noLimit` creates unbounded memory risk for JoinDiff — all 11 driver files

When `noLimit` is true, no LIMIT clause is added. For JoinDiff, the result set could be millions of rows loaded into Node.js memory. Consider a safety limit (e.g., 10M rows) even with `noLimit`.

### Nit (1)

- `simulation-suite.test.ts` changes (removing `data:` wrapper, flattening mock response fields) are unrelated bug fixes that should be in a separate commit

### Positive Observations

1. **Auto-discovery of audit columns** with two-layer exclusion (name patterns + schema default detection) is production-quality. The SKILL.md workflow requiring user confirmation before excluding is excellent UX
2. **CTE injection for query-vs-query comparison** is clever — wrapping arbitrary SQL in CTEs so the Rust engine treats them as tables
3. **Partition support** (date/numeric/categorical) with per-partition breakdown reporting makes this usable on billion-row tables
4. **SKILL.md** is thorough — the 9-step plan template, algorithm selection guide, `extra_columns` behavior docs, and common mistakes section are exactly what an LLM agent needs
5. **Consistent `noLimit` across all 11 drivers** — clean, mechanical refactor with correct `effectiveLimit > 0` guard for truncation detection
6. **Defensive null guards** added to `lineage-check.ts`, `schema-inspect.ts`, `sql-analyze.ts` — good hardening of existing tools

### Missing Tests

1. No test for `injectCte` merging engine CTEs with source/target CTEs
2. No test for partition discovery failure (error propagation from `runPartitionedDiff`)
3. No integration test exercising `DataParitySession` NAPI round-trip from TypeScript

---
*Reviewed by 1 model: Claude. External models (Gemini 3.1 Pro, Kimi K2.5, Grok 4) unavailable due to insufficient OpenRouter credits.*

🤖 Generated with [Claude Code](https://claude.com/claude-code)

**Review by suryaiyer95 — COMMENTED** [review]
## Multi-Model Code Review — PR #493 + PR #114

**Verdict: REQUEST CHANGES**
**Critical Issues: 1** | **Major Issues: 6** | **Minor Issues: 4** | **Nits: 3**

*Reviewed by 3 AI models: Claude, Gemini 3.1 Pro, Grok 4. 1 convergence round. Full review posted on companion PR altimate-core-internal #114.*

This PR's TypeScript orchestrator is well-crafted — the `noLimit` driver flag, CTE injection, and column auto-discovery are clean implementations. The issues below are specific to this PR.

---

### Issues in this PR

**`isQuery` detection could misclassify table names** (Minor — Logic Error)
- **File:** `packages/opencode/src/altimate/native/connections/data-diff.ts:23-26`
- Regex `/^\s*(SELECT|WITH|VALUES)\b/i` misclassifies table names like `with_metadata` or `select_results`.
- **Fix:** Document the limitation or require explicit signaling (e.g., a separate `sourceIsQuery` parameter).
- *Flagged by: Claude*

**Implicit truthy check on `effectiveLimit`** (Nit — Code Quality)
- **File:** `packages/drivers/src/*.ts` (all 12 driver files)
- The limit-appending condition checks `effectiveLimit && !/\bLIMIT\b/i.test(sql)`. Since `noLimit: true` sets `effectiveLimit` to `0` (falsy), this works but relies on implicit coercion.
- **Fix:** Change to `effectiveLimit > 0` for explicit intent.
- *Flagged by: Gemini*

---

### Positive Observations

- **`noLimit` flag** — Clean, consistent addition across all 10+ drivers with backwards-compatible optional parameter
- **Auto-timestamp exclusion** — Two-layer detection (name patterns + schema-level DEFAULT analysis) is smart
- **ClickHouse stderr silencing** — `clientConfig.log = { level: 127 }` prevents raw ERROR output from corrupting terminal TUI
- **CTE injection** — `injectCte` correctly handles merging CTE blocks when engine emits its own `WITH` clauses
- **Well-designed SKILL.md** — Critical sections on algorithm selection, cost-aware interaction, joindiff cross-DB trap

---

### Missing Tests

- No test for `isQuery` edge cases (table names like `with_metadata`)
- No test for CTE injection edge cases (nested/RECURSIVE CTEs)
- No partitioned diff simulation tests end-to-end
- No test for `discoverExtraColumns` with auto-timestamp default detection

---

*See full review with all 14 findings and attribution table on [altimate-core-internal PR #114](https://github.com/AltimateAI/altimate-core-internal/pull/114).*

File: packages/opencode/src/altimate/native/connections/data-diff.ts [hunk 1]
```diff
@@ -0,0 +1,879 @@
...
+    const lo = Number(partitionValue)
+    const hi = lo + bucketSize!
+    return `${quotedCol} >= ${lo} AND ${quotedCol} < ${hi}`
+  }
+
+  if (mode === "categorical") {
+    // Quote the value — works for strings, enums, booleans
+    const escaped = partitionValue.replace(/'/g, "''")
+    return `${quotedCol} = '${escaped}'`
+  }
```
[the remark this answers is not in the document]
Comment (suryaiyer95) [reply packages/opencode/src/altimate/native/connections/data-diff.ts]: Valid — the partition WHERE clause uses ANSI double-quotes which MySQL treats as string literals by default. The Rust engine handles dialect-specific quoting for its own SQL, but these partition predicates are generated in TypeScript. Will add dialect-aware quoting to `buildPartitionWhereClause`.

File: packages/opencode/src/altimate/native/connections/data-diff.ts [hunk 2]
```diff
@@ -0,0 +1,879 @@
...
+    case "bigquery":
+      return `${expr} = '${partitionValue}'`
+    case "clickhouse":
+      return `${expr} = toDate('${partitionValue}')`
+    case "mysql":
+    case "mariadb":
+      return `${expr} = '${partitionValue}'`
+    default:
+      return `${expr} = '${partitionValue}'`
+  }
```
[the remark this answers is not in the document]
Comment (suryaiyer95) [reply packages/opencode/src/altimate/native/connections/data-diff.ts]: Acknowledged — minor edge case. Partition discovery returns string representations from the database driver. For categorical columns the single-quote wrapping works for strings and numerics (databases auto-cast `'123'` to INT in comparisons). Boolean columns on BigQuery could be an issue but this is a rare combination (partitioning by boolean). Will add a note.

File: packages/opencode/src/altimate/native/connections/data-diff.ts [hunk 3]
```diff
@@ -0,0 +1,879 @@
...
+  const mode = partitionMode(granularity, bucketSize)
+
+  let expr: string
+  if (mode === "numeric") {
+    expr = `FLOOR(${partitionColumn} / ${bucketSize}) * ${bucketSize}`
+  } else if (mode === "date") {
+    expr = dateTruncExpr(granularity!, partitionColumn, dialect)
+  } else {
+    // categorical — raw distinct values, no transformation
+    expr = partitionColumn
```
[the remark this answers is not in the document]
Comment (suryaiyer95) [reply packages/opencode/src/altimate/native/connections/data-diff.ts]: Valid — `buildPartitionDiscoverySQL` should quote the partition column. Will fix alongside the `buildPartitionWhereClause` quoting fix to use consistent dialect-aware quoting.

File: packages/opencode/src/altimate/native/connections/data-diff.ts [hunk 4]
```diff
@@ -0,0 +1,879 @@
...
+      return `${expr} = '${partitionValue}'`
+    case "clickhouse":
+      return `${expr} = toDate('${partitionValue}')`
+    case "mysql":
+    case "mariadb":
+      return `${expr} = '${partitionValue}'`
+    default:
+      return `${expr} = '${partitionValue}'`
+  }
+}
```
[the remark this answers is not in the document]
Comment (suryaiyer95) [reply packages/opencode/src/altimate/native/connections/data-diff.ts]: Duplicate of earlier comment — see reply above.

File: packages/opencode/src/altimate/tools/data-diff.ts [hunk 5]
```diff
@@ -0,0 +1,257 @@
...
+      .describe("For numeric partition columns: size of each bucket. E.g. 100000 splits l_orderkey into ranges of 100K. Omit for date or categorical columns."),
+  }),
+  async execute(args, ctx) {
+    // Require read permission — data diff executes SELECT queries
+    await ctx.ask({
+      permission: "sql_execute_read",
+      patterns: [args.source.slice(0, 120), args.target.slice(0, 120)],
+      always: ["*"],
+      metadata: {},
+    })
```
[the remark this answers is not in the document]
Comment (suryaiyer95) [reply packages/opencode/src/altimate/tools/data-diff.ts]: Not applicable — `sql_execute_read` is the correct permission. The data_diff tool only executes SELECT queries (via the Rust engine's cooperative protocol). No DDL/DML is ever executed. The permission check exists to surface the tool's database access to the user for approval.
